File: Dec2020/9Dec/SumOfLeftLeaves.cpp
/*
Problem name: 404. Sum of Left Leaves
Problem link: https://leetcode.com/problems/sum-of-left-leaves/
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        int val = 0;
        
        if(!root)
            return 0;
        if(root->left && !root->left->left && !root->left->right)
            val = root -> left ->val;
        
        return val + sumOfLeftLeaves(root->left) + sumOfLeftLeaves(root->right);
    }
};

File: Dec2020/9Dec/SumOfLeftLeaves.java
/*
Problem Name: 404. Sum of Left Leaves
Problem Link: https://leetcode.com/problems/sum-of-left-leaves/
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int sum;
    public int sumOfLeftLeaves(TreeNode root) {
        sum = 0;
        traverse(root,false);
        return sum;
    }
    void traverse(TreeNode root,boolean isleft){
        if(root == null)
            return;
        if((root.left == null) && (root.right == null) && isleft)
            sum += root.val;
        traverse(root.left,true);
        traverse(root.right,false);
    }
}
